application/src/control.rs

```diff
@@ -0,0 +1,387 @@
+                            xfer.accept_with(&cfg.json).ok();
+                        }
+                        ConfigKey::UsbConsole => {
+                            xfer.accept_with(&cfg.usb_console).ok();
+                        }
+                        ConfigKey::PowerOn => {
+                            xfer.accept_with(&cfg.power_on).ok();
+                        }
+                        ConfigKey::PowerOff => {
+                            xfer.accept_with(&cfg.power_off).ok();
+                        }
+                        ConfigKey::PowerRescue => {
+                            xfer.accept_with(&cfg.power_rescue).ok();
+                        }
+                    }
+                } else {
+                    xfer.reject().unwrap();
+                }
+            }
+            Ok(ControlRequest::Read) => {
+                if let Ok(key) = req.value.try_into() {
+                    match key {
+                        ReadKey::Version => {
+                            let mut buf = heapless::Vec::<u8, 128>::new();
+                            crate::version::write_version(&mut buf);
+                            xfer.accept_with(&buf).ok();
+                        }
+                        ReadKey::Power => {
+                            let mut buf = heapless::Vec::<u8, 128>::new();
+                            write!(buf, "{:.2}W", self.data.power).ok();
+                            xfer.accept_with(&buf).ok();
+                        }
+                        ReadKey::Voltage => {
+                            let mut buf = heapless::Vec::<u8, 128>::new();
+                            write!(buf, "{:.2}V", self.data.voltage).ok();
+                            xfer.accept_with(&buf).ok();
+                        }
+                        ReadKey::Current => {
+                            let mut buf = heapless::Vec::<u8, 128>::new();
+                            write!(buf, "{:.2}A", self.data.current).ok();
+                            xfer.accept_with(&buf).ok();
+                        }
+                    }
+                } else {
+                    xfer.reject().unwrap();
+                }
+            }
+            _ => {
+                xfer.reject().unwrap();
+            }
+        }
+    }
+
+    fn control_out(&mut self, xfer: ControlOut<B>) {
+        let req = xfer.request();
+
+        match req {
+            &Request {
+                request_type: RequestType::Vendor,
+                recipient: Recipient::Interface,
+                index,
+                ..
+            } if index as u8 == self.iface.into() => (),
+            _ => return,
+        }
+
+        match req.request.try_into() {
+            Ok(ControlRequest::Nop) => {
+                xfer.accept().unwrap();
+            }
+            Ok(ControlRequest::Power) => {
+                if let Ok(action) = req.value.try_into() {
+                    self.power = Some(action);
```

Comment:
ah nice, so you decouple the reception of the control commands/data, with the actuator. I guess it isolates the objects each task has access to.